### PR TITLE
#3398 #3399 #3400

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
#3399 edit(file = '') is broken on VS 2017
- not specific to 2017.
- Fixed incorrect calculation of ~/
- Made reliable opening of files since some 'traditional' methods fail on CPS
- If file is already opened, just activate

#3398 Get tooltip for nested function call
- Use AST rather than signature help utilities

#3400 Update package installed in Program Files crashed RTVS
- Hide button as we do with uninstall. Packages in Program Files location cannot be uninstalled or updated
- Added exception handling just in case

### CODE REVIEW

File: src/Package/Impl/DataInspect/Viewers/FileEditor.cs [hunk 1]
```diff
@@ -58,31 +58,32 @@ internal sealed class FileEditor : IFileEditor {
 
             if (!string.IsNullOrEmpty(fileName)) {
                 try {
-                    if (!Path.IsPathRooted(fileName)) {
+                    if (fileName.StartsWithOrdinal("~\\")) {
```
Comment (int19h) [inline src/Package/Impl/DataInspect/Viewers/FileEditor.cs]: Shouldn't this (also) check for `~/`?

Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/Viewers/FileEditor.cs]: It is converted `fileName?.FromRPath();` above

File: src/Common/Core/Impl/Extensions/IOExtensions.cs [hunk 2]
```diff
@@ -135,5 +135,15 @@ public static class IOExtensions {
             }
             return allowWrite && !denyWrite;
         }
+
+        public static string NormalizePath(this string path) {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/IOExtensions.cs]: How is it different from `System.IO.Path.GetFullPath`?

Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Extensions/IOExtensions.cs]: Does `GetFullPath` normalize? I thought it does nothing to rooted ie `c:\foo\.\..\` remains. But if it indeed does the right thing, we dont need `NormalizePath`.

Comment (AlexanderSher) [inline src/Common/Core/Impl/Extensions/IOExtensions.cs]: At least `System.IO.Path.GetFullPath(@"c:\FusionLog/..\NVIDIA/DisplayDriver\")` produced `"c:\\NVIDIA\\DisplayDriver\\"`

File: src/R/Editor/Test/QuickInfo/QuickInfoSourceTest.cs [hunk 3]
```diff
@@ -66,16 +65,26 @@ class Session {
                 .Which.ToString().Should().StartWith("as.Date.character(x, ...)");
         }
 
+        [CompositeTest]
+        [InlineData(2, "sin(")]
+        [InlineData(5, "cos(")]
+        [InlineData(8, "cos(")]
+        [InlineData(11, "sin(")]
+        public async Task NestedCallTest(int position, string name) {
```
Comment (AlexanderSher) [inline src/R/Editor/Test/QuickInfo/QuickInfoSourceTest.cs]: Do you want to check for second nested parameter and for multiple nested?

File: src/R/Editor/Impl/QuickInfo/QuickInfoSource.cs [hunk 4]
```diff
@@ -57,61 +59,82 @@ internal sealed class QuickInfoSource : IQuickInfoSource {
...
+                int wrapLength = Math.Min(SignatureInfo.MaxSignatureLength, signatureString.Length);
+                string text;
+
+                if (string.IsNullOrWhiteSpace(functionInfo.Description)) {
+                    text = string.Empty;
+                } else {
+                    // VS may end showing very long tooltip so we need to keep 
+                    // description reasonably short: typically about
+                    // same length as the function signature.
+                    text = signatureString + "\r\n" + functionInfo.Description.Wrap(wrapLength);
```
Comment (AlexanderSher) [inline src/R/Editor/Impl/QuickInfo/QuickInfoSource.cs]: `FormattableString`?

Comment (MikhailArkhipov) [inline src/R/Editor/Impl/QuickInfo/QuickInfoSource.cs]: Yeah, I will look into cleaning this up in the editor split. There is bunch of old code here.

File: src/R/Editor/Test/QuickInfo/QuickInfoSourceTest.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/R/Editor/Test/QuickInfo/QuickInfoSourceTest.cs]: Sure we can do that

File: src/Common/Core/Impl/Extensions/IOExtensions.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/Common/Core/Impl/Extensions/IOExtensions.cs]: OK.

**Comment by AlexanderSher** [comment]
👍 
@int19h , can you please take a look one more time?

**Comment by int19h** [comment]
👍

**Pull request merged**